### PR TITLE
Add unit test for sessions repository

### PR DESCRIPTION
**Pull request by sys1yagi**
## Issue

Add test for SessionsRepository #96

## Overview (Required)

- Add SessionsRepositoryTest and a first test for hasCacheSessions(). 	27ce6ac
- Move subscribeOn and observeOn calls from the data source to where we actually use each Observable. 102e6a9
- Add tests for `SessionsRepository.findAll()`. 102e6a9
- Add hasSession(int) and test.			5ef2fd2
- Fix SessionsRemoteDataSource.find(int, String).  d36bfae
- Add tests for find(int, String). bf8d36d

## Links

None

## Screenshot

None

## Todo

- [x] Discuss about Observable thread control
- [x] Add test for `find()`
- [x] update description
- [x] fix local cache test.

### CODE REVIEW

File: app/src/main/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsLocalDataSource.java [hunk 1]
```diff
@@ -45,9 +45,7 @@ private Topic_Relation topicRelation() {
 
     @Override
     public Single<List<Session>> findAll(String languageId) {
-        return sessionRelation().selector().executeAsObservable().toList()
-                .subscribeOn(Schedulers.io())
-                .observeOn(AndroidSchedulers.mainThread());
```
Comment (sys1yagi) [inline app/src/main/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsLocalDataSource.java]: IMO: I think that the thread control should be decided by actual use place such as Activity, Fragment, Presenter etc.  Also, can't test Observable containing thread operation with Observable.test().

**Comment by konifar** [comment]
:eyes:
To be honest, I'm not very familiar with testing 🙇 
So I'll learn from this and discuss.

File: app/src/main/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsLocalDataSource.java (hunk 1 above)
Comment (konifar) [inline app/src/main/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsLocalDataSource.java]: > Also, can't test Observable containing thread operation with Observable.test().

Make sense. I'm for this change.

Comment (kgmyshin) [inline app/src/main/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsLocalDataSource.java]: > I think that the thread control should be decided by actual use place such as Activity, Fragment, Presenter etc.

I think so too 🎅

Comment (sys1yagi) [inline app/src/main/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsLocalDataSource.java]: thanks~

File: app/src/main/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRemoteDataSource.java [hunk 2]
```diff
@@ -35,17 +37,16 @@ public SessionsRemoteDataSource(DroidKaigiClient client) {
...
     @Override
     public Maybe<Session> find(int sessionId, String languageId) {
-        return findAll(languageId).map(sessions -> Stream.of(sessions)
+        return findAll(languageId)
+                .toObservable()
+                .flatMap(Observable::fromIterable)
                 .filter(session -> session.id == sessionId)
-                .findSingle()
-                .get()).toMaybe();
+                .singleElement();
```
Comment (sys1yagi) [inline app/src/main/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRemoteDataSource.java]: It crashes if there is no value in the original implementation. So I replaced it with flatMap.

**Comment by sys1yagi** [comment]
Please review. 🙏

**Comment by sys1yagi** [comment]
I have found a bug due to this change...
Please wait a moment as I fix it.

**Comment by konifar** [comment]
Sorry, conflict occurred.. Can you merge master?

**Comment by sys1yagi** [comment]
yes~

File: app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        run {
+            val repository = SessionsRepository(
+                    SessionsLocalDataSource(mock()),
+                    SessionsRemoteDataSource(mock())
+            )
+            repository.cachedSessions = mapOf(0 to Session())
+            repository.setIdDirty(false)
+
+            assertThat(repository.hasCacheSessions()).isTrue()
+        }
```
Comment (jmatsu) [inline app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt]: Could you please add a case that `repository has any cached session, but repository is dirty`?

File: app/src/main/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsLocalDataSource.java (hunk 1 above)
Comment (amay077) [inline app/src/main/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsLocalDataSource.java]: > I think that the thread control should be decided by actual use place such as Activity, Fragment, Presenter etc.

In MVVM, ViewModel changes scheduler to mainThread.

File: app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+        // create cache
+        repository.findAll(Session.LANG_JA_ID)
+                .test()
+                .run{
+                    assertNoErrors()
+                }
+
+        repository.setIdDirty(true)
```
Comment (jmatsu) [inline app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt]: It seems this case will check wether or not the repository can find the session which is associated with id `12` from *remote* source.

Maybe `setIdDirty(false)` is correct? 
If so, `cachedSessions.verify(never()).get(eq(12))` should be also modified to `cachedSessions.verify().get(eq(12))`

File: app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt (hunk 3 above)
Comment (sys1yagi) [inline app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt]: ok~

File: app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+import io.reactivex.Single
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mockito.*
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.RuntimeEnvironment
+
+@RunWith(RobolectricTestRunner::class)
+class SessionsRepositoryTest {
```
Comment (KeithYokoma) [inline app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt]: Sweet! 😍

**Comment by KeithYokoma** [comment]
LGTM!

File: app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt (hunk 4 above)
Comment (sys1yagi) [inline app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt]: nice catch 👏  I intended to call LocalDataSource. However, the code calls RemoteDataSource.

Comment (sys1yagi) [inline app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt]: fix it 502b5b16aa87a2c14242b69e6a04d71834b043bd

File: app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt (hunk 3 above)
Comment (sys1yagi) [inline app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt]: add test 1c03487f15e4da08543c11f5adb2be3c07614507

**Comment by jmatsu** [comment]
👀

File: app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt (hunk 4 above)
Comment (jmatsu) [inline app/src/test/java/io/github/droidkaigi/confsched2017/repository/sessions/SessionsRepositoryTest.kt]: 😱  I noticed that `If so, cachedSessions.verify(never()).get(eq(12)) should be also modified to cachedSessions.verify().get(eq(12))` is incorrect review... Sorry for confusing you 🙇 
Anyway, thank you for fixing!

**Comment by jmatsu** [comment]
LGTM!

**Comment by konifar** [comment]
Thank you so much! 👏  💪

**Pull request merged**